server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remoteDelivery/DelaysAndMaxRetry.java

```diff
@@ -31,71 +28,70 @@
-
-        // Check consistency of 'maxRetries' with delayTimesList attempts.
+        List<Delay> delayTimesList = createDelayList(delaysAsString);
         int totalAttempts = calcTotalAttempts(delayTimesList);
+        return getDelaysAndMaxRetry(intendedMaxRetries, totalAttempts, delayTimesList);
+    }
 
-        // If inconsistency found, fix it.
+    private static DelaysAndMaxRetry getDelaysAndMaxRetry(int intendedMaxRetries, int totalAttempts, List<Delay> delayTimesList) throws MessagingException {
         if (totalAttempts > intendedMaxRetries) {
-            LOGGER.warn("Total number of delayTime attempts exceeds maxRetries specified. " + " Increasing maxRetries from " + intendedMaxRetries + " to " + totalAttempts);
+            LOGGER.warn("Total number of delayTime attempts exceeds maxRetries specified. Increasing maxRetries from " + intendedMaxRetries + " to " + totalAttempts);
             return new DelaysAndMaxRetry(totalAttempts, delayTimesList);
         } else {
             int extra = intendedMaxRetries - totalAttempts;
-            if (extra != 0) {
-                LOGGER.warn("maxRetries is larger than total number of attempts specified.  " + "Increasing last delayTime with " + extra + " attempts ");
-
-                // Add extra attempts to the last delayTime.
-                if (delayTimesList.size() != 0) {
-                    // Get the last delayTime.
-                    Delay delay = delayTimesList.get(delayTimesList.size() - 1);
-
-                    // Increase no. of attempts.
-                    delay.setAttempts(delay.getAttempts() + extra);
-                    LOGGER.warn("Delay of " + delay.getDelayTime() + " msecs is now attempted: " + delay.getAttempts() + " times");
-                } else {
-                    throw new MessagingException("No delaytimes, cannot continue");
-                }
+            if (extra > 0) {
+                LOGGER.warn("maxRetries is larger than total number of attempts specified. Increasing last delayTime with " + extra + " attempts ");
+                return addExtraAttemptToLastDelay(intendedMaxRetries, extra, delayTimesList);
             }
             return new DelaysAndMaxRetry(intendedMaxRetries, delayTimesList);
         }
     }
 
-    private static ArrayList<Delay> createDelayList(String delaysAsString) {
-        ArrayList<Delay> delayTimesList = new ArrayList<Delay>();
-        try {
-            if (delaysAsString != null) {
+    private static DelaysAndMaxRetry addExtraAttemptToLastDelay(int intendedMaxRetries, int extra, List<Delay> delayTimesList) throws MessagingException {
+        if (delayTimesList.size() != 0) {
+            Delay lastDelay = delayTimesList.get(delayTimesList.size() - 1);
+            LOGGER.warn("Delay of " + lastDelay.getDelayTimeInMs() + " msecs is now attempted: " + lastDelay.getAttempts() + " times");
+            return new DelaysAndMaxRetry(intendedMaxRetries,
```

Comment:
You can work with Iterables.limit and Iterables.concat instead of creating real lists